### PR TITLE
[JSON]Add sort video items by original title - SortByOriginalTitle

### PR DESCRIPTION
**Pull request by JimmyS83**
This is Matrix version of a (now closed) PR originally submitted for Leia - https://github.com/xbmc/xbmc/pull/19434

## Description
Adds sort method, according `originaltitle` variable/infolabel, with fallback to standard `sorttitle`, `title`.

## Motivation and Context
Most of (newer) scrapers (such as default tmdb) supports scraping item `Title` in user’s choice language. They are able to store also original title into `originaltitle` variable (infolabel).

It could be displayed by skin in lists, etc. User can choose, which title he wants in lists, dynamically, even both of them for example.

The problem is, that Kodi lacks function to sort list according originaltitle. It supports only `Title`, and `Sorttitle` parameters. `Sorttitle` is written to DB during scraping, so its not dynamic parameter at all.

## Screenshots (if appropriate):
Example on dummy data, where both titles are visible to make a point more easilly: https://streamable.com/kgvjj9

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [X] I have added tests to cover my change
- [X] All new and existing tests passed

### CODE REVIEW

**Comment by fuzzard** [comment]
You will actually need to PR this against master branch.
Our process is anything new must be merged into master, and then a backport may be considered for the previous version after this.
Having said that, we generally don't backport features, only fixes, but I'll let @DaveTBlake be the decider on this one

**Comment by wsnipex** [comment]
jenkins build this please

**Comment by JimmyS83** [comment]
any update?

File: xbmc/interfaces/json-rpc/schema/types.json
Comment (Montellese) [inline xbmc/interfaces/json-rpc/schema/types.json:112]: You also need to bump the minor part of the JSON API version.

And please fix the commit message. It should at least mention `JSON-RPC` and `sort by original title`.

**Review by Montellese — CHANGES_REQUESTED** [review]
The changes look good apart from the missing JSON-RPC API version bump.

And please fix your commit messages to use at least a sentence like `add support to sort items by original title` or something similar.

File: xbmc/interfaces/json-rpc/schema/types.json
Comment (JimmyS83) [reply xbmc/interfaces/json-rpc/schema/types.json:112]: Better now?

Comment (Montellese) [reply xbmc/interfaces/json-rpc/schema/types.json:112]: Yes as long as you squash the three commits concerning JSON-RPC into one :+1:

Comment (JimmyS83) [reply xbmc/interfaces/json-rpc/schema/types.json:112]: Done :)

**Comment by DaveTBlake** [comment]
I think that the first commit could also do with a better description too. For example: 
"Add sorting of video items by original title".

File: xbmc/filesystem/PluginDirectory.cpp
Comment (DaveTBlake) [inline xbmc/filesystem/PluginDirectory.cpp:244]: We are working towards making all code to the same standard so new lines of code needs to be to our current style guides rather than copy the style around it. This looks way too long. Run `clang` on the changed/added lines of code.

Comment (JimmyS83) [reply xbmc/filesystem/PluginDirectory.cpp:244]: I dont understand. I followed existing format of a file which I changed (SORT_METHOD_VIDEO_SORT_TITLE_IGNORE_THE for reference). This is minor change, not refactor bigger chunks of application like filesystem part. I dont have such intentions..

Comment (Montellese) [reply xbmc/filesystem/PluginDirectory.cpp:244]: Just wait for the jenkins build to finish and it will automatically leave a comment here with the formatting changes necessary. Then you can apply those to your commits.

**Review by Montellese — APPROVED** [review]
Looks good to me, thanks.

@DaveTBlake do you want to hold this back as it poses the same issue as #19372 concerning JSON-RPC version bumps?

**Comment by DaveTBlake** [comment]
> do you want to hold this back as it poses the same issue as #19372 concerning JSON-RPC version bumps?

Yes, would be useful to hold merging this until after 19.2 point release it out.

**Review by Montellese — CHANGES_REQUESTED** [review]
Looks good in general but I found a few things that need to be addressed. It also looks like you lost the JSON-RPC version bump in your rebase.

**Review by Montellese — CHANGES_REQUESTED** [review]
Apart from the version bump which needs fixing everything is fine for me now. Thanks for your patience.

**Review by Montellese — APPROVED** [review]

**Comment by phunkyfish** [comment]
@JimmyS83 you appear to have lost the version bump again.

**Comment by JimmyS83** [comment]
@phunkyfish Hello, yes, I started to be a bit tired of updating just JSON number over and over, when this doesnt seems to be any closer to the merge.. :(

**Comment by phunkyfish** [comment]
> @phunkyfish Hello, yes, I started to be a bit tired of updating just JSON number over and over, when this doesnt seems to be any closer to the merge.. :(

Fair point. I’ve get that pain. Let’s see can we get this merged.

@DaveTBlake you ok with this being merged once the version is added?

File: addons/resource.language.en_gb/resources/strings.po
Comment (ksooo) [inline addons/resource.language.en_gb/resources/strings.po:13316]: Please add a verbal description to help our translators so they can get their job done.
I mean thise lines starting with `#.` ;-)

File: xbmc/utils/SortUtils.cpp
Comment (ksooo) [inline xbmc/utils/SortUtils.cpp:289]: Maybe `if (!title.empty() && (attributes & ...))` ?

Comment (JimmyS83) [reply xbmc/utils/SortUtils.cpp:289]: Are you sure that _SortAttributeIgnoreArticle_ should be dependent on non-empty title? In that case I think _std::string BySortTitle_ must be changed also for consistency

File: addons/resource.language.en_gb/resources/strings.po
Comment (JimmyS83) [reply addons/resource.language.en_gb/resources/strings.po:13317]: done

File: addons/resource.language.en_gb/resources/strings.po
Comment (ksooo) [inline addons/resource.language.en_gb/resources/strings.po]: nope. the line must start with #. (hashmark, followed by dot)
And I'm honestly not sure this description is actually helpful. You should at least explain what an "original title" is, so a translator gets an idea *what* this string denotes.

**Review by ksooo — CHANGES_REQUESTED** [review]

File: xbmc/utils/SortUtils.cpp
Comment (ksooo) [reply xbmc/utils/SortUtils.cpp:289]: what could the result of `SortUtils::RemoveArticles` called with an empty string be? nothing else tan an ampty string, right? So, if title is empty, there should be no need to call `SortUtils::RemoveArticles`. That's the whole idea.

Comment (Montellese) [reply xbmc/utils/SortUtils.cpp:289]: I don't think `title` will ever be empty because we have the fallback to use `FieldTitle` and I don't think many people have items without titles in their libraries.

File: addons/resource.language.en_gb/resources/strings.po
Comment (Montellese) [reply addons/resource.language.en_gb/resources/strings.po]: The term "original title" has been there for a long time and this PR just cleans up the usage comments. I don't see why he has to add a detailed description of what "original title" means.

File: xbmc/utils/SortUtils.cpp
Comment (ksooo) [reply xbmc/utils/SortUtils.cpp:289]: Okay, then, Thanks for clarifying, @Montellese

File: addons/resource.language.en_gb/resources/strings.po
Comment (ksooo) [reply addons/resource.language.en_gb/resources/strings.po]: Translators in Weblate may not be familar with our code to look up the details there. "Original title" is quite a generic term - so, a translator might ask: original title of what? Of a movie? A PVR recording? That might make a difference in non-english languages. A comment explaining what this string is actually used for is really very helpful - says somebody who translated many hundreds of strings for Kodi, skins and add-ons.

Comment (JimmyS83) [reply addons/resource.language.en_gb/resources/strings.po]: #. Original title of the movie/tvshow 

is fine for you?

I am struggling with better definition of original, as this term is known for me since years (original title vs "country distributor title" which could differ or not )..
image

**Comment by JimmyS83** [comment]
> > @phunkyfish Hello, yes, I started to be a bit tired of updating just JSON number over and over, when this doesnt seems to be any closer to the merge.. :(
> 
> Fair point. I’ve get that pain. Let’s see can we get this merged.
> 
> @DaveTBlake you ok with this being merged once the version is added?

I am fine with bump number of JSON. Current is I believe 12.7.0. If somebody really merge this after resolving all other possible things, just tell me wanted version number, I will add it before squash/rebase at the end.

**Comment by JimmyS83** [comment]
I added requested comment for translators, I added again JSON version (which was changed again 2 days ago), but I dont know if anyone merge this before another JSON version bump...

**Review by ksooo — APPROVED** [review]

**Comment by ksooo** [comment]
@JimmyS83 please fix the clang-format issues, otherwise the PR wont get merged.

**Comment by JimmyS83** [comment]
> @JimmyS83 please fix the clang-format issues, otherwise the PR wont get merged.

I added formatting changes proposed by jenkins.

**Comment by ksooo** [comment]
> I added formatting changes proposed by jenkins.

 Almost. ;-) Jenkins is still not happy.

**Comment by lrusak** [comment]
jenkins build this please

**Comment by lrusak** [comment]
jenkins build this please

**Comment by JimmyS83** [comment]
I think that one machine has problems with disc space or smth, otherwise everything should be OK.

**Comment by phunkyfish** [comment]
Jenkins build this please

**Review by phunkyfish — APPROVED** [review]

**Comment by JimmyS83** [comment]
Merge me please :-)

**Comment by phunkyfish** [comment]
Thanks for sticking with it @JimmyS83!

**Pull request merged**
